### PR TITLE
Milestones!

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:**

## Feature Details
Adds Milestones. A Milestone is defined in the game data with conditions to complete it.
Milestones also support states beyond just 'complete' and 'incomplete'.
The intention is for Milestones to appear in a special tab of hte logbook panel.

## UI Screenshots
Haven't got there yet.

## Usage Examples
{{if this feature is used in the data files, provide examples!}}

## Testing Done
{{describe how you tested the new feature}}

## Performance Impact
{{describe any performance impact (positive or negative). "N/A" if no performance-critical code is changed. }}

### CODE REVIEW

**Comment by Hurleveur** [comment]
This is really something I would love to see in the game; maybe we could fuse it up with the way the tutorial system works currently; exporting that into a new class that checks player conditions and other things to know when to show milestones or tutorial prompts?
Probably for later, knowing how it works currently... The tutorial prompt are a bit place randomly throughout the code, just something I thought of, like a system to check if currently the player could get a tutorial/milestone with his condition and current menu etc

**Comment by Fzzr** [comment]
> This is really something I would love to see in the game; maybe we could fuse it up with the way the tutorial system works currently; exporting that into a new class that checks player conditions and other things to know when to show milestones or tutorial prompts?
> Probably for later, knowing how it works currently... The tutorial prompt are a bit place randomly throughout the code, just something I thought of, like a system to check if currently the player could get a tutorial/milestone with his condition and current menu etc

I think the current plan mostly covers that. All the states will be controlled by conditions just like any other content (missions etc.) Milestones have these states:

- `hidden` - the milestone is completely invisible, not in the list to the player at all. This is meant to be used for milestones that are spoilers, like the fact that there are multiple endings to the main storyline. Some of these will become gradually revealed, but fully secret milestones that are spoilers until they are achieved like say visiting Sagittarius A* would go straight from `hidden` to `complete`. 
- `locked` - the milestone is visible, but not achievable. This won't actually stop people from writing code to transition straight from `locked` to `complete`, but it allows partially hidden states such as obfuscating the description and de-emphasizing in the UI. This would be for things that are considered to be guaranteed to be reached but not spoilers for existing. For example, there might be a locked milestone "Make three friends on the other side of the rainbow" about meeting the three main factions only available with the jump drive that's locked until you actually get the jump drive.
- `unlocked` - Visible and achievable. It's still possible for mildly spoilery milestones to have obfuscated descriptions here. For example, a milestone to meet the Hai is technically achievable from the start of the game, but the description should probably be minimally descriptive. 
- `complete` - Just what you'd think, this is the final state. Unlike the other states, by default this one will probably give the player a prompt of some kind. 

Some of the things being planned are intended for use by plugins rather than the base game, like prompts for transitioning to locked or unlocked. Likewise I think in the base game milestones should be considered to be at the leaf of condition logic. The reason not to check milestones in other conditions is it leads to things being a bit too meta. Since milestones are already based on conditions, basing content on the status of milestones just adds another layer of indirection to figuring out how logic flows, plus as milestones can have multiple ways to transition between states they're inherently less stable for the purposes of controlling flow.

File: source/Milestone.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/* Milestone.cpp
+Copyright (c) 2022 by warp-core
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/Milestone.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

File: source/Milestone.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/* Milestone.h
+Copyright (c) 2022 by warp-core
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/Milestone.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

File: source/milestonestate.hpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
+/* milestonestate.hpp
+Copyright (c) 2022 by warp-core
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/milestonestate.hpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

**Comment by dahe-at-google** [comment]
Just noticed this. It seems to play well the the storylines proposal in https://github.com/endless-sky/endless-sky/issues/7449.

**Comment by Amazinite** [comment]
Closing this PR since it has been in the draft phase for some time without much recent activity. You can reopen it once you've finalized your changes and it's ready for review.

**Pull request closed without merging**